v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,465 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import asyncio
+import functools
+import logging
+import paho.mqtt.client as mqtt
+import weakref
+
+
+logger = logging.getLogger(__name__)
+
+OP_TYPE_CONNECT = "CONNECT"
+OP_TYPE_DISCONNECT = "DISCONNECT"
+
+RECONNECT_MODE_DROP = "RECONNECT_DROP"
+RECONNECT_MODE_ALL = "RECONNECT_ALL"
+
+
+class MQTTError(Exception):
+    """Represents a failure with a Paho-given error rc code"""
+
+    def __init__(self, rc):
+        self.rc = rc
+        super().__init__(mqtt.error_string(rc))
+
+
+class MQTTConnectionFailedError(Exception):
+    """Represents a failure to connect.
+    Can have a Paho-given connack rc code, or a message"""
+
+    def __init__(self, rc=None, message=None, fatal=False):
+        if not rc and not message:
+            raise ValueError("must provide rc or message")
+        if rc and message:
+            raise ValueError("rc and message are mutually exclusive")
+        self.rc = rc
+        self.fatal = fatal
+        if rc:
+            message = mqtt.connack_string(rc)
+        super().__init__(message)
+
+
+class ConnectionLock(asyncio.Lock):
+    """Async Lock with additional attributes corresponding to the operation.
```

Comment:
> """Async Lock with additional attributes corresponding to the operation. [](http://example.com/codeflow?start=3&length=73)

I'm having trouble reasoning about the behavior of this object. It's like a critical section or a mutex, but it also contains the result of that operation (the future), but the result is only valid until the operation is complete (when the lock is released). So it's like a lock and an "on connection completed" object combined in one., except the "on connection completed" object is really limited in when you can use it.